cmd/background-newdisks-heal-ops.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"time"
+
+	"github.com/minio/minio/cmd/logger"
+)
+
+const defaultMonitorNewDiskInterval = time.Minute * 10
+
+func initLocalDisksAutoHeal() {
+	go monitorLocalDisksAndHeal()
+}
+
+// monitorLocalDisksAndHeal - ensures that detected new disks are healed
+//  1. Only the concerned erasure set will be listed and healed
+//  2. Only the node hosting the disk is responsible to perform the heal
+func monitorLocalDisksAndHeal() {
+	// Wait until the object layer is ready
+	var objAPI ObjectLayer
+	for {
+		objAPI = newObjectLayerFn()
+		if objAPI == nil {
+			time.Sleep(time.Second)
+			continue
+		}
+		break
+	}
+
+	sets, ok := objAPI.(*xlSets)
+	if !ok {
+		return
+	}
+
+	ctx := context.Background()
+
+	var bgSeq *healSequence
+	var found bool
+
+	for {
+		bgSeq, found = globalBackgroundHealState.getHealSequenceByToken(bgHealingUUID)
+		if found {
+			break
+		}
+		time.Sleep(time.Second)
+	}
+
+	// Perform automatic disk healing when a new one is inserted
+	for {
+		time.Sleep(defaultMonitorNewDiskInterval)
+
+		localDisksToHeal := []Endpoint{}
+		for _, endpoint := range globalEndpoints {
+			if !endpoint.IsLocal {
+				continue
+			}
+			// Try to connect to the current endpoint
+			// and reformat if the current disk is not formatted
+			_, _, err := connectEndpoint(endpoint)
+			if err == errUnformattedDisk {
+				localDisksToHeal = append(localDisksToHeal, endpoint)
+			}
+		}
+
+		if len(localDisksToHeal) == 0 {
+			continue
+		}
+
+		// Reformat disks
+		bgSeq.sourceCh <- SlashSeparator
+		// Ensure that reformatting disks is finished
+		bgSeq.sourceCh <- nopHeal
+
+		// Compute the list of erasure set to heal
+		var erasureSetToHeal []int
+		for _, endpoint := range localDisksToHeal {
+			// Load the new format of this passed endpoint
+			_, format, err := connectEndpoint(endpoint)
```

Comment:
> Totally right, I missed that. It should be fixed, the new code calculates the list of erasure sets that need to be healed so more than new disk can be healed in a single operation.

Looks like this is not fixed yet and you should do the following? :
```
--- a/cmd/background-newdisks-heal-ops.go
+++ b/cmd/background-newdisks-heal-ops.go
@@ -89,7 +89,7 @@ func monitorLocalDisksAndHeal() {
                bgSeq.sourceCh <- nopHeal
 
                // Compute the list of erasure set to heal
-               var erasureSetToHeal []int
+               erasureSetToHeal := make([]bool, len(sets.sets))
                for _, endpoint := range localDisksToHeal {
                        // Load the new format of this passed endpoint
                        _, format, err := connectEndpoint(endpoint)
@@ -104,11 +104,14 @@ func monitorLocalDisksAndHeal() {
                                continue
                        }
 
-                       erasureSetToHeal = append(erasureSetToHeal, setIndex)
+                       erasureSetToHeal[setIndex] = true
                }
 
                // Heal all erasure sets that need
-               for _, setIndex := range erasureSetToHeal {
+               for setIndex, heal := range erasureSetToHeal {
+                       if !heal {
+                               continue
+                       }
                        xlObj := sets.sets[setIndex]
                        err := healErasureSet(ctx, setIndex, xlObj)
                        if err != nil {
```

Reply:
> Looks like this is not fixed yet and you should do the following? :

Your code works but also my original code should also work as expected, why do you think this is not fixed yet ?

Reply:
> erasureSetToHeal = append(erasureSetToHeal, setIndex)

since you are doing append, same erasure set number can get appended (for different new disks) causing healErasureSet() to be called twice on the same erasureset

Reply:
> since you are doing append, same erasure set number can get appended (for different new disks) causing healErasureSet() to be called twice on the same erasureset

oups yes.. fixed!